cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java

```diff
@@ -426,6 +435,26 @@ public void onSuccess(T result, int statusCode, Map<String, List<String>> respon
    * @return a {@link CompletionStage} that will complete when the delete operation completed
    */
   private CompletionStage<String> deleteResource(int gracePeriodSeconds) {
+    if (isUserProgram) {
+      // Delete per-run configmap
+      String configMapName = KubeTwillPreparer.CONFIGMAP_NAME_PREFIX + runId.getId();
+      try {
+        coreV1Api.deleteNamespacedConfigMap(configMapName, kubeNamespace, null, null, null, null,
+                                            null, null);
+      } catch (ApiException e) {
+        if (e.getCode() == 404) {
+          LOG.trace("Configmap {} doesn't exist for resource {} of type {}",
+                    configMapName, meta.getName(), resourceType);
+        } else {
+          LOG.warn("Failed to delete configmap {} for resource {} of type {}. Error code: {}, body: {}",
+                   configMapName, meta.getName(), resourceType, e.getCode(), e.getResponseBody());
+          CompletableFuture<String> resultFuture = new CompletableFuture<>();
+          completeExceptionally(resultFuture, e);
```

Comment:
what happens to the program state if this happens? Does it transition to a failed state? Does the deletion get retried?

Reply:
Program state transitions to failed